### PR TITLE
Inbox/pagination — part 1

### PR DESCRIPTION
**Pull request by NelsonVides**
Currently inbox introduces only the max key of the [RSM XEP](https://xmpp.org/extensions/xep-0059.html), and within the inbox form we also allow for setting something equivalent to the RSM’s before and after. But there’s no RSM payload in the iq-fin-result, and also, helper libraries might be used on the client side for more complex pagination, provided full support for RSM is available.

This PR introduces support for points:
- [Limiting the Number of Items](https://xmpp.org/extensions/xep-0059.html#limit)
- [Paging Forwards Through a Result Set](https://xmpp.org/extensions/xep-0059.html#forwards)
- [Paging Backwards Through a Result Set](https://xmpp.org/extensions/xep-0059.html#backwards)

Other points in the document might be implemented fully or partially but I haven't taken them into consideration for this PR and are therefore not verified nor tested.

Note that the `count` is not introduced, the document specifies that:

> The responding entity SHOULD also include the number of items in the full result set (which MAY be approximate) encapsulated in a <count/> element. The <first/> element SHOULD include an 'index' attribute. This integer specifies the position within the full set (which MAY be approximate) of the first item in the page. If that item is the first in the full set, then the index SHOULD be '0'. If the last item in the page is the last item in the full set, then the value of the <first/> element's 'index' attribute SHOULD be the specified count minus the number of items in the last page.

> Note: The <count/> element and 'index' attribute enable important functionality for requesting entities (for example, a scroll-bar user-interface component). They MAY be omitted, but only if it would be either impossible or exceptionally resource intensive to calculate reasonably accurate values.

Also not yet fully perfected is the fact that the UIDs as specified in the document are not unique, currently it is using the integer representation of the timestamp of the inbox row, which can potentially collide. Left is to ensure timestamps are unique, or use a hash of the jid which are known to be unique, though in that case it'd need to make more complex queries to find out about the order by timestamp.

This is for now experimental and therefore not yet fully documented, I plan to do so in a subsequent PR, but it is as of this PR pretty much the requirement I currently have.

### CODE REVIEW

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/147296) / small_tests / 503c73c5e21e80188cce92a0272fef62485446f6
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3827/147296/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3827/147296/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-10-26_12.46.12/lib.mongooseim.logs/run.2022-10-26_12.46.12/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/148530) / small_tests / f2121536b7181e5a08c890179956015a0e2e6999
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3827/148530/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3827/148530/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-11-03_05.20.47/lib.mongooseim.logs/run.2022-11-03_05.20.48/suite.log.html)

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/148773) / small_tests / a9d463e0cfa5d4cecf1efff55b3324da7be9f321
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3827/148773/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3827/148773/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-11-03_14.29.40/lib.mongooseim.logs/run.2022-11-03_14.29.41/suite.log.html)

File: src/inbox/mod_inbox.erl [hunk 1]
```diff
@@ -430,36 +449,39 @@ text_single_form_field(Var, DefaultValue) ->
...
+build_params(Params, #rsm_in{max = Max, id = undefined}) when Max =/= undefined ->
+    Params#{limit => Max};
+build_params(Params0, Rsm) ->
+    Params = maps:without([start, 'end'], Params0),
+    build_params_with_rsm(Params#{rsm => Rsm}, Rsm).
+
+build_params_with_rsm(Params, #rsm_in{max = Max, id = <<>>, direction = before}) ->
+    Params#{limit => Max, order => asc, start => 0};
+build_params_with_rsm(Params, #rsm_in{max = Max, id = <<>>, direction = aft}) ->
+    Params#{limit => Max, 'end' => (1 bsl 63 - 1)};
```
Comment (chrzaszcz) [inline src/inbox/mod_inbox.erl:475]: Why do we have to pass an integer here? If we skip it, wouldn't the DB query just omit this unnecessary guard parameter?

**Review by chrzaszcz — COMMENTED** [review]
Looks good, I added a few minor comments.

**Comment by mongoose-im** [comment]
[small_tests_24](https://circleci.com/gh/esl/MongooseIM/150195) / small_tests / 5d3c85d4137b063ee837179080164d639a7b59ce
Reports [root](http://esl.github.io/circleci-mim-results/s3_reports.html?prefix=PR/3827/150195/small_tests.24.3.4) / [small](https://circleci-mim-results.s3.eu-central-1.amazonaws.com/PR/3827/150195/small_tests.24.3.4/small/ct_run.mongooseim@localhost.2022-11-08_12.57.07/lib.mongooseim.logs/run.2022-11-08_12.57.08/suite.log.html)

**Pull request merged**